---
title: "CaCTS Tutorial"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

**Ca**ncer **C**ore **T**ranscription factor **S**pecificity (CaCTS) is an algorithm to identify candidate MTFs using pan-cancer RNA-sequencing data from The Cancer Genome Atlas. In this page we describe the procedures to perform CaCTS analysis. For more results details, please check our [biorixv](https://www.biorxiv.org/content/10.1101/839142v1) preprint .

# 1. Installation
Install all packages in the latest version of [R](https://www.r-project.org/).

### Option 1
Downaload [CaCTS](https://github.com/lawrenson-lab/CaCTS_zip/raw/master/CaCTS_1.0.tar.gz) package and run the following command in a *Terminal*

```{r package, echo=TRUE, eval=FALSE}
R CMD INSTALL CaCTS_1.0.tar.gz
```

### Option 2
First, install the devtools package.

```{r install_devtools, echo=TRUE, eval=FALSE}
install.packages("devtools")
```
Load the devtools package.
```{r devtools, echo=TRUE, eval=FALSE}
library(devtools)
```
Use install_github as follows:
Load the devtools package.
```{r install_github, echo=TRUE, eval=FALSE}
install_github("lawrenson-lab/CaCTS_pkg")
```
Load the CaCTS package.
```{r CaCTS}
library(CaCTS)
```

## Pre-processing input files

Basically, CaCTS algorithm requires two main data structure inputs: the *expression matrix* and the *annotation table*.

```{r annotation}

# Exoression data
load("data/TCGA.RNA.Rda")
TCGA.RNA[1:4, 1:4]
# Annotations
annot.table <- read.delim("data/SuppTable1-34-TCGAID.txt")
colnames(annot.table)[3] <- "sample.id"
colnames(annot.table)[2] <- "group.name"
head(annot.table)

annot.table$group.name <- paste0(annot.table$Cancer, "-", annot.table$group.name)
head(data.frame(table(annot.table$group.name)), 20)
```

In this step, from all Pancancer expressed genes, we are selecting only TF genes. Published lists of transcription factors (TF) were retrieved from Saint-André et al., 2016 and Lambert et al., 2018. Merging both lists we created a catalogue of 1,671 unique TFs, of which 1,578 were expressed in the pancan data set.

```{r filtering}
TF.list = read.delim("data/merged.list.1671.TFs.txt", sep = "\t")

f.TCGA.RNA = TCGA.RNA[which(TCGA.RNA$gene_symbol %in% as.character(TF.list$NameTF)),]
dim(f.TCGA.RNA)

aux = which(!as.character(TF.list$NameTF) %in% f.TCGA.RNA$gene_symbol)
write.table(TF.list[aux,],file = "data/non-expressed-TFs.txt", quote = F, row.names = F)

rownames(f.TCGA.RNA) <- f.TCGA.RNA$gene_symbol
f.TCGA.RNA <- f.TCGA.RNA[,3:ncol(f.TCGA.RNA)]
f.TCGA.RNA[1:5, 1:4]
dim(f.TCGA.RNA)

```
To calculate a Jensen-Shannon Divergence (JSD) score for each TF for each tumor type, we adjusted the normalized expression values, to rescale all values to >0.

```{r reschale, eval=TRUE}
delta1 = max(f.TCGA.RNA, na.rm = T) - min(f.TCGA.RNA, na.rm = T)
delta2 = max(f.TCGA.RNA, na.rm = T) - 0

f.TCGA.RNA.rs = (f.TCGA.RNA - min(f.TCGA.RNA, na.rm = T)) * delta1 / delta2
f.TCGA.RNA.rs[1:5, 1:4]
dim(f.TCGA.RNA.rs)
```

## Step 1 - Representative sample

Next, for each tumor type, a representative sample was generated by calculating the mean expression value for each TF.

```{r prepare, echo=TRUE, eval=TRUE}
matrix.rep <- prepare_representaive_samples(expr.matrix = f.TCGA.RNA.rs, sample.descr = annot.table, save.file = F)
matrix.rep[1:4, 1:4]
```

## Step 2 - Calculatiog the JSD score

```{r loop, echo=TRUE, message = FALSE, warning = FALSE, eval=TRUE}
cancer = "OV-OV"
res.CaCTS <- run_CaCTS_score(matrix.rep, cancer)
```

```{r visualize, echo=TRUE, message = FALSE, warning = FALSE, eval=TRUE}
tail(res.CaCTS)
p = visualize_scores(tf.scores = res.CaCTS, topn = 0, ncol = 1, rep.matrix = matrix.rep, query.name = cancer, filename = paste0(cancer, "-CaCTS-scores.pdf"), w=5, h=5)
p
```

The final candidate MTF list for each cancer type was defined by considering the intersection of the 5% most highly expressed TFs in each tumor type, and the top 5% TFs when ranked by JSD score.

```{r filter candidates}

filtered <- filter_by_expression_rank(rep.matrix = matrix.rep, tf.scores = res.CaCTS, query.name = cancer, pn=0.05, pnE=0.05)
filtered

```

To predict the candidate MTFs for the other cancer types, the same procedures on **Step  2** can be performed by changin the cancer of interest.

## References

D’Alessio, A.C., Fan, Z.P., Wert, K.J., Baranov, P., Cohen, M.A., Saini, J.S., Cohick, E., Charniga, C., Dadon, D., Hannett, N.M., et al. (2015). A Systematic Approach to Identify Candidate Transcription Factors that Control Cell Identity. Stem Cell Reports 5, 763–775.

Lambert, S.A., Jolma, A., Campitelli, L.F., Das, P.K., Yin, Y., Albu, M., Chen, X., Taipale, J., Hughes, T.R., and Weirauch, M.T. (2018). The human transcription factors. Cell 175, 598–599.

Saint-André, V., Federation, A.J., Lin, C.Y., Abraham, B.J., Reddy, J., Lee, T.I., Bradner, J.E., and Young, R.A. (2016). Models of human core transcriptional regulatory circuitries. Genome Res. 26, 385–396.